File: shared-utils/src/main/java/com/pretrade/utils/TimeUtils.java
package com.pretrade.utils;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.time.DayOfWeek;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.temporal.TemporalAdjusters;

/**
 * Utility class for IST (Indian Standard Time) time operations
 * relevant to NSE market hours and expiry calculations.
 */
@Slf4j
@Component
public class TimeUtils {

    private static final ZoneId IST_ZONE = ZoneId.of("Asia/Kolkata");

    private static final LocalTime PRE_MARKET_OPEN = LocalTime.of(9, 0);
    private static final LocalTime MARKET_OPEN = LocalTime.of(9, 15);
    private static final LocalTime MARKET_CLOSE = LocalTime.of(15, 30);

    /**
     * Returns the current date-time in IST.
     *
     * @return current LocalDateTime in Asia/Kolkata timezone
     */
    public LocalDateTime nowIST() {
        return LocalDateTime.now(IST_ZONE);
    }

    /**
     * Checks if the current time falls within the pre-market window (9:00 AM - 9:15 AM IST).
     *
     * @return true if currently in pre-market session
     */
    public boolean isPreMarketWindow() {
        LocalDateTime now = nowIST();
        LocalTime currentTime = now.toLocalTime();
        return isWeekday(now.toLocalDate())
                && !currentTime.isBefore(PRE_MARKET_OPEN)
                && currentTime.isBefore(MARKET_OPEN);
    }

    /**
     * Checks if the market is currently open (9:15 AM - 3:30 PM IST on weekdays).
     *
     * @return true if the market is currently in regular trading session
     */
    public boolean isMarketOpen() {
        LocalDateTime now = nowIST();
        LocalTime currentTime = now.toLocalTime();
        return isWeekday(now.toLocalDate())
                && !currentTime.isBefore(MARKET_OPEN)
                && !currentTime.isAfter(MARKET_CLOSE);
    }

    /**
     * Checks if the market is currently closed.
     * The market is closed outside 9:00 AM - 3:30 PM IST or on weekends.
     *
     * @return true if the market is closed
     */
    public boolean isMarketClosed() {
        return !isPreMarketWindow() && !isMarketOpen();
    }

    /**
     * Calculates the current weekly expiry date (Thursday).
     * If today is past Thursday, returns next Thursday.
     * If today is Thursday and market is closed, returns next Thursday.
     *
     * @return the current expiry date as LocalDate
     */
    public LocalDate getCurrentExpiry() {
        LocalDateTime now = nowIST();
        LocalDate today = now.toLocalDate();
        DayOfWeek dayOfWeek = today.getDayOfWeek();

        // Find this week's Thursday
        LocalDate thisThursday = today.with(TemporalAdjusters.nextOrSame(DayOfWeek.THURSDAY));

        // If today is Thursday and market is already closed, move to next Thursday
        if (dayOfWeek == DayOfWeek.THURSDAY && now.toLocalTime().isAfter(MARKET_CLOSE)) {
            return thisThursday.plusWeeks(1);
        }

        // If we've already passed Thursday this week, get next Thursday
        if (dayOfWeek.getValue() > DayOfWeek.THURSDAY.getValue()) {
            return today.with(TemporalAdjusters.next(DayOfWeek.THURSDAY));
        }

        return thisThursday;
    }

    /**
     * Calculates the next weekly expiry date (the Thursday after the current expiry).
     *
     * @return the next expiry date as LocalDate
     */
    public LocalDate getNextExpiry() {
        return getCurrentExpiry().plusWeeks(1);
    }

    /**
     * Converts an Instant to IST LocalDateTime.
     *
     * @param instant the Instant to convert
     * @return the corresponding LocalDateTime in IST
     */
    public LocalDateTime toIST(Instant instant) {
        if (instant == null) {
            return null;
        }
        return LocalDateTime.ofInstant(instant, IST_ZONE);
    }

    /**
     * Checks if today (IST) is a weekday (Monday-Friday).
     *
     * @return true if today is a weekday
     */
    public boolean isWeekday() {
        return isWeekday(nowIST().toLocalDate());
    }

    /**
     * Checks if a given date is a weekday (Monday-Friday).
     *
     * @param date the date to check
     * @return true if the date is a weekday
     */
    private boolean isWeekday(LocalDate date) {
        DayOfWeek day = date.getDayOfWeek();
        return day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY;
    }

    /**
     * Returns the IST ZoneId for external use.
     *
     * @return the Asia/Kolkata ZoneId
     */
    public ZoneId getISTZone() {
        return IST_ZONE;
    }

    /**
     * Returns the current ZonedDateTime in IST.
     *
     * @return current ZonedDateTime in Asia/Kolkata timezone
     */
    public ZonedDateTime nowZonedIST() {
        return ZonedDateTime.now(IST_ZONE);
    }
}

File: shared/time_utils.py
"""IST (Indian Standard Time) utilities for NSE market hours and expiry calculations.

All time operations in the system MUST go through this module.
Never use datetime.now() directly — always use now_ist().
"""

from datetime import datetime, date, time, timedelta
from zoneinfo import ZoneInfo

IST = ZoneInfo("Asia/Kolkata")

PRE_MARKET_OPEN = time(9, 0)
MARKET_OPEN = time(9, 15)
MARKET_CLOSE = time(15, 30)


def now_ist() -> datetime:
    """Returns the current datetime in IST."""
    return datetime.now(IST)


def today_ist() -> date:
    """Returns today's date in IST."""
    return now_ist().date()


def is_pre_market_window() -> bool:
    """Checks if current time is in the pre-market window (9:00-9:15 AM IST on weekdays)."""
    now = now_ist()
    t = now.time()
    return is_weekday(now.date()) and PRE_MARKET_OPEN <= t < MARKET_OPEN


def is_market_open() -> bool:
    """Checks if the market is currently open (9:15 AM - 3:30 PM IST on weekdays)."""
    now = now_ist()
    t = now.time()
    return is_weekday(now.date()) and MARKET_OPEN <= t <= MARKET_CLOSE


def is_market_closed() -> bool:
    """Checks if the market is currently closed."""
    return not is_pre_market_window() and not is_market_open()


def is_weekday(d: date = None) -> bool:
    """Checks if a date is a weekday (Monday-Friday)."""
    if d is None:
        d = today_ist()
    return d.weekday() < 5


def get_current_expiry() -> date:
    """Calculates the current weekly expiry date (Thursday).

    If today is past Thursday, returns next Thursday.
    If today is Thursday and market is closed, returns next Thursday.
    """
    now = now_ist()
    today = now.date()
    days_until_thursday = (3 - today.weekday()) % 7
    thursday = today + timedelta(days=days_until_thursday)

    if today == thursday and now.time() > MARKET_CLOSE:
        thursday += timedelta(days=7)
    if today.weekday() > 3:
        thursday = today + timedelta(days=(3 - today.weekday()) % 7 + 7)

    return thursday


def get_next_expiry() -> date:
    """Returns the next weekly expiry (Thursday after current expiry)."""
    return get_current_expiry() + timedelta(days=7)


def format_timestamp(dt: datetime) -> str:
    """Formats a datetime as 'dd-MMM-yyyy HH:mm:ss'."""
    if dt is None:
        return "N/A"
    return dt.strftime("%d-%b-%Y %H:%M:%S")
